Add Window Controls Overlay article (#4680)
GoogleChrome/web.dev: src/site/content/en/blog/window-controls-overlay/index.md

| 5. Launch                                | Not started              |

</div>

### Enabling via chrome://flags

To experiment with Window Controls Overlay locally, without an origin trial token, enable the
`#enable-desktop-pwas-window-controls-overlay` flag in `chrome://flags`.

### Enabling support during the origin trial phase

Starting in Chrome&nbsp;92, Window Controls Overlay will be available as an origin trial in Chrome.
The origin trial is expected to end in Chrome&nbsp;94 (expected in July 2021).

{% include 'content/origin-trials.njk' %}

### Register for the origin trial {: #register-for-ot }

{% include 'content/origin-trial-register.njk' %}

## How to use Window Controls Overlay

### Adding `window-controls-overlay` to the Web App Manifest

A progressive web app can opt-in to the window controls overlay by adding
`"window-controls-overlay"` as the primary `"display_override"` member in the web app manifest:

```json
{
  "display_override": ["window-controls-overlay"]
}
```

The window controls overlay will be visible only when all of the following conditions are satisfied:

1. The app is _not_ opened in the browser, but in a separate PWA window.

> 1. The manifest includes `"display_override": ["window-controls-overlay"]"`. (Other values are

Reviewer: Trailing " too much

Author: Oh, good one. Fixed.

> 1. The manifest includes `"display_override": ["window-controls-overlay"]"`. (Other values are

Author: ```suggestion
1. The manifest includes `"display_override": ["window-controls-overlay"]`. (Other values are
```

> 1. The manifest includes `"display_override": ["window-controls-overlay"]"`. (Other values are

Reviewer 2: I'm sorry I didn't find this before.

> 1. The manifest includes `"display_override": ["window-controls-overlay"]"`. (Other values are

Author: Fixed already! Thanks!

  position: fixed;
  /**
   * Gradient, because why not. Endless opportunities.
   * The gradient ends in maroon, which happens to be the app's
   * `<meta name="theme-color" content="maroon">`.
   */
  background-image: linear-gradient(90deg, #131313, 33%, maroon);
  /* Use the environment variable for the left anchoring with a fallback. */
  left: env(titlebar-area-x, 0);
  /* Use the environment variable for the top anchoring with a fallback. */
  top: env(titlebar-area-y, 0);
  /* Use the environment variable for setting the width with a fallback. */
  width: env(titlebar-area-width, 100%);
  /* Use the environment variable for setting the height with a fallback. */
  height: env(titlebar-area-height, 33px);
}
```

You can see the effect of this code in the screenshot below. The title bar is fully responsive. When
you resize the PWA window, the title bar reacts as if it were composed of regular HTML content,
which, in fact, it is.

{% Img src="image/8WbTDNrhLsU0El80frMBGE4eMCD3/5dc2j3CfrKczvTaASvKE.png", alt="App window with a search bar in the title bar.", width="800", height="112" %}

### Determining which parts of the title bar are draggable

> While the screenshot above suggests that you are done, you are not done quite yet. The PWA window is

Reviewer 3: Although it hasn't yet been implemented, there will be a portion of the overlay that is draggable. This protects the user's ability to drag the window in case the app fails to implement a draggable region. My interpretation of this sentence is that the entire overlay is not draggable, which isn't quite true--although _most_ of the overlay will not be draggable.

Author: Gotcha. I tried every pixel, but makes sense it didn't work since it isn't implemented yet.

no longer draggable, since the window controls buttons are no drag areas, and the rest of the title

> bar consists of the search widget. This can be fixed by leveraging the `app-region` CSS property

Reviewer 3: `app-region` hasn't been standardized yet, so the plan was to continue using the webkit prefixed `-webkit-app-region` until `app-region` is standardized. Currently, only `-webkit-app-region` is supported in the browser.

Author: I add this as an aside.

`windowControlsOverlay`:

```js
if ('windowControlsOverlay' in navigator) {
  // Window Controls Overlay is supported.
}
```

### Querying the window controls region with `windowControlsOverlay`

The code so far has only one problem: on some platforms the window controls are on the right, on
others they are on the left. This means that the linear gradient background image needs to be
dynamically adapted to run from `#131313` to `maroon` or vice versa, so that it blends in with the
title bar's `maroon` background color that is determined by
`<meta name="theme-color" content="maroon">`. This can be achieved by querying the
[`getBoundingClientRect()`](https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect)
API on the `navigator.windowControlsOverlay` property.

```js
if ('windowControlsOverlay' in navigator) {
  const { x } = navigator.windowControlsOverlay.getBoundingClientRect();
  // Window controls are on the right (like on Windows).
  // [ windowControlsOverlay___________________ [_] [■] [X] ]
  if (x === 0) {
    div.classList.add('search-controls-right');
  }
  // Window controls are on the left (like on macOS).

>   // [ [X] [_] [■] ___________________windowControlsOverlay ]

Author: Yuck, I wasn't aware of this. This makes https://github.com/WICG/window-controls-overlay/issues/14 even more desirable.

  else {
    div.classList.add('search-controls-left');
  }
} else {
  // When running in a non-supporting browser tab.
  div.classList.add('search-controls-right');
}
```

Rather than having the background image in the `.search` class CSS rules directly (as before), the
modified code now uses two classes that the code above sets dynamically.

```css
.search-controls-left {
  background-image: linear-gradient(90deg, maroon, 66%, #131313);
}

.search-controls-right {
  background-image: linear-gradient(90deg, #131313, 33%, maroon);
}
```

### Determining if the window controls overlay is visible

The window controls overlay will not be visible in the title bar area in all circumstances. While it
will naturally not be there on browsers that do not support the Window Controls Overlay feature, it
will also not be there when the PWA in question runs in a tab. To detect this situation, you can
query the `visible` property of the `windowControlsOverlay`:

```js
if (navigator.windowControlsOverlay.visible) {
  // The window controls overlay is visible in the title bar area.
}
```

> {% Aside %} The window controls overlay visibility is not to be confused with visibility in the CSS

Reviewer 3: The way I read this:
> Even if you set `display: none` on the window controls overlay, the `visible` property would
still report `true`.

It seems like it seems to suggest that a dev could modify the window controls overlay, but the overlay is a piece of browser UI that cannot be modified by the webcontent. By "window controls overlay" are you referring to the pseudo-titlebar div that's next to the overlay?

Author: I'm referring to… 

```html
<div class="search">
  <img src="chuck-norris.png" alt="Chuck Norris" width="32" height="32" />
  <label>
    <input type="search" />
    Search words in jokes
  </label>
</div>
```

…in my example. I need to phrase it like the `div` with the search bar that gets displayed in the window controls overlay.

sense. Even if you set `display: none` on the window controls overlay, the `visible` property would
still report `true`. {% endAside %}

### Being notified of geometry changes

Querying the window controls overlay area with `getBoundingClientRect()` can suffice for one-off
things like setting the correct background image based on where the window controls are, but in
other cases, more fine-grained control is necessary. For example, a possible use case could be to
adapt the window controls overlay based on the available space and to add a joke right in the window
control overlay when there is enough space.

{% Img src="image/8WbTDNrhLsU0El80frMBGE4eMCD3/Unm5S2dB3szfFF2YCcFO.png", alt="Window controls overlay area on wide window displaying an additional joke (which reads Chuck Norris can find a word that rhymes with orange.)", width="800", height="73" %}

You can be notified of geometry changes by subscribing to

> `navigator.windowControlsOverlay.ongemometrychange` or by setting up an event listener for the

Reviewer: It's ongeometry

> `navigator.windowControlsOverlay.ongemometrychange` or by setting up an event listener for the

Author: ```suggestion
`navigator.windowControlsOverlay.ongeometrychange` or by setting up an event listener for the
```

      clearTimeout(timeout);
      func(...args);
    };
    clearTimeout(timeout);
    timeout = setTimeout(later, wait);
  };
};

if ('windowControlsOverlay' in navigator) {
  navigator.windowControlsOverlay.ongeometrychange = debounce((e) => {
    span.hidden = e.boundingRect.width < 800;
  }, 250);
}
```

Rather than assigning a function to `ongeometrychange`, you can also add an event listener to
`windowControlsOverlay` like below. You can read up on the difference between the two on
[MDN](https://developer.mozilla.org/en-US/docs/Web/Guide/Events/Event_handlers).

```js
navigator.windowControlsOverlay.addEventListener(
  'geometrychange',
  debounce((e) => {
    span.hidden = e.boundingRect.width < 800;
  }, 250),
);
```

### Compatibility when running in a tab and on non-supporting browsers

There are two possible cases to consider:

- The case where an app is running in a browser that _does_ support Window Controls Overlay, but
  where the app is used in a browser tab.
- The case where an app is running in a browser that _does not_ support Window Controls Overlay.

In both cases, by default the window controls overlay will display inline like regular HTML content

> and the `env()` variables' fallback values will kick in. On supporting browsers, you can also decide

Reviewer 3: > On supporting browsers, you can also decide to not display the window controls overlay by checking the overlay's `visible` property and then hiding it.

This isn't something the developer would control using the readonly `visible` property. Instead, the toggle that is mentioned in the security section is intended for the user. It allows them to start off in a safe state with a `standalone` title bar, then they can toggle the WCO on using a button in the title bar and toggle it back off if they want to return back to a `standalone` window.

Author: I have made this clearer by separating more clearly between the window controls overlay and the HTML designated to be placed inside of the window controls overlay.

---
layout: post
title: Customize the window controls overlay of your PWA's title bar
subhead: |
  Make use of the title bar area next to the window controls to make your PWA feel more like an app.
authors:
  - thomassteiner
  - amandabaker

> date: 2021-04-20

Author: ```suggestion
date: 2021-04-22
```

`windowControlsOverlay`:

```js
if ('windowControlsOverlay' in navigator) {
  // Window Controls Overlay is supported.
}
```

### Querying the window controls region with `windowControlsOverlay`

The code so far has only one problem: on some platforms the window controls are on the right, on
others they are on the left. To make matters worse, the "three dots" Chrome menu will change
position, too, based on the platform. This means that the linear gradient background image needs to
be dynamically adapted to run from `#131313`→`maroon` or `maroon`→`#131313`→`maroon`, so that it
blends in with the title bar's `maroon` background color that is determined by
`<meta name="theme-color" content="maroon">`. This can be achieved by querying the
[`getBoundingClientRect()`](https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect)
API on the `navigator.windowControlsOverlay` property.

```js
if ('windowControlsOverlay' in navigator) {
  const { x } = navigator.windowControlsOverlay.getBoundingClientRect();
  // Window controls are on the right (like on Windows).
  // Chrome menu is left of the window controls.

>   // [ windowControlsOverlay___________________ […] [_] [■] [X] ]

Author: It's actually meant to reflect how it looks like on Windows (horizontal …) vs. on macOS (vertical ⋮).

  }
} else {
  // When running in a non-supporting browser tab.
  div.classList.add('search-controls-right');
}
```

Rather than having the background image in the `.search` class CSS rules directly (as before), the
modified code now uses two classes that the code above sets dynamically.

```css
/* For macOS: */
.search-controls-left {
  background-image: linear-gradient(90deg, maroon, 45%, #131313, 90%, maroon);
}

/* For Windows: */
.search-controls-right {
  background-image: linear-gradient(90deg, #131313, 33%, maroon);
}
```

### Determining if the window controls overlay is visible

The window controls overlay will not be visible in the title bar area in all circumstances. While it
will naturally not be there on browsers that do not support the Window Controls Overlay feature, it
will also not be there when the PWA in question runs in a tab. To detect this situation, you can
query the `visible` property of the `windowControlsOverlay`:

```js
if (navigator.windowControlsOverlay.visible) {
  // The window controls overlay is visible in the title bar area.
}
```

{% Aside %} The window controls overlay visibility is not to be confused with the visibility in the

> CSS sense of whatever HTML content you place in the window controls overlay. Even if you set

Author: ```suggestion
CSS sense of whatever HTML content you place next to the window controls overlay. Even if you set
```

> CSS sense of whatever HTML content you place in the window controls overlay. Even if you set

Author: Understood. I have changed this to "next to".

### Enabling support during the origin trial phase

Starting in Chrome&nbsp;92, Window Controls Overlay will be available as an origin trial in Chrome.

> The origin trial is expected to end in Chrome&nbsp;94 (July 2021).

Reviewer 2: ```suggestion
The origin trial is expected to end in Chrome&nbsp;94 (expected in July 2021).
```
